### PR TITLE
Finish custom type implementation

### PR DESCRIPTION
**Pull request by Lipt0nas**
It is/was a bit of a messy implementation, due to ordering of imports and declarations needing to be in a more strict order than other languages

### CODE REVIEW

File: bindgen/src/bindings/cpp/templates/wrapper.hpp [hunk 1]
```diff
@@ -42,6 +61,19 @@ struct {{ name }};
...
 struct {{ name }};
+{%- when Type::Custom { module_path, name, builtin } %}
+{%- match config.custom_types.get(name.as_str()) %}
+{%- when None %}
+typedef {{ builtin|type_name }} {{ name }};
+{%- when Some with (type_config) %}
+{%- match type_config.type_name %}
+{%- when Some with (type_name) %}
+typedef {{ type_name }} {{ name }};
+{%- let type_name = typ|type_name %}
```
Comment (arg0d) [inline bindgen/src/bindings/cpp/templates/wrapper.hpp:72]: Does `type_name` and `class_name` actually do anything here?

File: bindgen/src/bindings/cpp/templates/wrapper.hpp [hunk 2]
```diff
@@ -14,6 +14,25 @@
 #include <type_traits>
 #include <variant>
 
+{%- for typ in ci.iter_types() %}
```
Comment (arg0d) [inline bindgen/src/bindings/cpp/templates/wrapper.hpp:17]: This looks like it should be in Rust code somewhere. Is the order of custom type imports important? From my experience with C# generator `add_include` sorts the includes, so adding this here "before" the include statements would not influence the order.

**Review by arg0d — APPROVED** [review]
Looks good overall!

File: bindgen/src/bindings/cpp/templates/wrapper.hpp (hunk 1 above)
Comment (Lipt0nas) [reply bindgen/src/bindings/cpp/templates/wrapper.hpp]: nice catch! left it by accident when moving code around, removed

File: bindgen/src/bindings/cpp/templates/wrapper.hpp (hunk 2 above)
Comment (Lipt0nas) [reply bindgen/src/bindings/cpp/templates/wrapper.hpp]: i totally forgot that you can also do things outside of templates :D, moved the import code to the wrapper generator constructor

**Comment by arg0d** [comment]
:+1: :+1:

**Pull request merged**